Mathlib/Tactic/Linter/FlexibleLinter.lean

```diff
@@ -0,0 +1,384 @@
+output of a "flexible" tactic (such as `simp`).
+
+For example, this ensures that, if you want to use `simp [...]` in the middle of a proof,
+then you should replace `simp [...]` by
+* a `suffices \"expr after simp\" by simpa` line;
+* the output of `simp? [...]`, so that the final code contains `simp only [...]`;
+* something else that does not involve `simp`!
+
+Otherwise, the linter will complain.
+
+## TODO
+The example
+```lean
+example (h : 0 = 0) : True := by
+  simp at h
+  assumption
+```
+should trigger the linter, since `assumption` uses `h` that has been "stained" by `simp at h`.
+However, `assumption` contains no syntax information for the location `h`, so the linter in its
+current form does not catch this.
+
+##  Implementation detail
+
+A large part of the code is devoted to tracking `FVar`s and `MVar`s between tactics.
+
+For the `FVar`s, this follows the following heuristic:
+* if the unique name of the `FVar` is preserved, then we use that;
+* otherwise, if the `userName` of the `FVar` is preserved, then we use that;
+* if neither is preserved, we drop the ball and stop tracking the `FVarId`.
+
+For the `MVar`s, we use the information of `Lean.Elab.TacticInfo.goalsBefore` and
+`Lean.Elab.TacticInfo.goalsAfter`.
+By looking at the `mvar`s that are either only "before" or only "after", we focus on the
+"active" goals.
+We then propagate all the `FVarId`s that were present in the "before" goals to the "after" goals,
+while leaving untouched the ones in the "inert" goals.
+-/
+
+open Lean Elab
+
+namespace Mathlib.Linter
+
+/-- The flexible linter makes sure that "rigid" tactics do not follow "flexible" tactics. -/
+register_option linter.flexible : Bool := {
+  defValue := false
+  descr := "enable the flexible linter"
+}
+
+namespace flexible
+
+/-- `flexible? stx` is `true` on syntax that takes a "wide" variety of inputs and modifies
+them in possibly unpredictable ways.
+
+The prototypical flexible tactic is `simp`
+The prototypical non-flexible tactic `rw`.
+`simp only` is also non-flexible. -/
+def flexible? : Syntax → Bool
+  | .node _ ``Lean.Parser.Tactic.simp #[_, _, _, only?, _, _] => only?[0].getAtomVal != "only"
+  | .node _ ``Lean.Parser.Tactic.simpAll #[_, _, _, only?, _] => only?[0].getAtomVal != "only"
+  | _ => false
+
+end flexible
+
+end Mathlib.Linter
+
+section goals_heuristic
+namespace Lean.Elab.TacticInfo
+
+/-!
+###  Heuristics for determining active and inactive goals
+
+The two definitions `activeGoalsBefore`, `activeGoalsAfter` extract a list of
+`MVarId`s attempting to determine which on which goals the tactic `t` is acting.
+This is mostly based on the heuristic that the tactic with "change" an `MVarId`.
+-/
+
+/-- `activeGoalsBefore t` are the `MVarId`s before the `TacticInfo` `t` that "disappear" after it.
+They should correspond to the goals in which the tactic `t` performs some action. -/
+def activeGoalsBefore (t : TacticInfo) : List MVarId :=
```

Comment:
Naming is hard: the current names sound to me like they describe *all* active goals before resp. after `t` (including the unchanged ones). I'm still thinking if I can think of better names...

Reply:
`goalsResolvedBefore` or `goalsClosedBefore` maybe? (what's the correct term to eradicate a goal?)

Reply:
I went for `goalsTargetedBy` and `goalsCreatedBy`, relating on the `TacticInfo` input to fill in `tactic`!